File: Soham_Petkar.jl
# using StrRegex

function parser(expr)
    expr = replace(expr, "^" => "")
    expr = replace(expr, " " => "")
    # terms = matchall(r"([+-]?\d*)([a-z]+\d*[a-z]?\d*)?", expr)
    terms = collect(eachmatch(r"([+-]?\d*)([a-z]+\d*[a-z]?\d*)?", expr))
    #println(terms)
    parsed_terms = []

    for term in terms
        #println(term)
        a, b = term
        #println(a)
        #println(b)
        coefficient_str = a
        if coefficient_str == "+"
            coefficient_str = "1"
        elseif coefficient_str == "-"
            coefficient_str = "-1"
        end

        if (isempty(coefficient_str))
            continue
        end
        coefficient = parse(Int, coefficient_str)
        #println(coefficient)
        variable = term[2]
        if isnothing(variable)
            variable = "const"
        end
        push!(parsed_terms, [coefficient, variable])
    end
    return parsed_terms
end

# Example usage
#expr = "4xy^2 + x^2y - 5x^2y^3 + 3"
#result = parser(expr)
#for i in result
#println(i)
#end

function add_polynomials(polynomials)
    combined_terms = Dict{String,Int}()
    for polynomial in polynomials
        terms = parser(polynomial)
        #println(terms)
        for term in terms
            coefficient, variable = term
            if haskey(combined_terms, variable)
                combined_terms[variable] += coefficient
            else
                combined_terms[variable] = coefficient
            end
        end
    end

    result = [(coefficient, variable)
              for (variable, coefficient) in pairs(combined_terms)]
    return result
end

# Example usage
polynomials = ["4xy^2 + x^2y - 5x^2y^3 +3", "2xy + 3x^2 - xy^2"]
result = add_polynomials(polynomials)
println(result)


# Example usage
function extract_columns(matrix)
    n = length(matrix)
    m = length(matrix[1])
    columns = Vector{Vector{AbstractString}}(undef, m)
    for j in 1:m
        column = Vector{AbstractString}(undef, n)
        for i in 1:n
            println(matrix[i][j])
            column[i] = matrix[i][j]
        end
        columns[j] = column
    end
    return columns
end

function extract_rows(matrix)
    rows = copy(matrix)
    return rows
end

function extract_diagonals(matrix)
    n = length(matrix)
    diagonals = Vector{Vector{AbstractString}}(undef, 2)
    diagonal = Vector{AbstractString}(undef, n)

    # Extract main diagonal
    for i in 1:n
        diagonal[i] = matrix[i][i]
    end
    push!(diagonals, diagonal)

    diagonal = Vector{AbstractString}(undef, n)
    # Extract secondary diagonal
    for i in 1:n
        diagonal[i] = matrix[i][n-i+1]
    end
    push!(diagonals, diagonal)

    return diagonals[3:end]
end


function extract_all(matrix)
    all_items = vcat(extract_columns(matrix), extract_rows(matrix), extract_diagonals(matrix))
    return all_items
end

# Example usage
##result = extract_all(matrix)
#println(result)

function magic(result)
    globalsum = add_polynomials(result[1])
    for item in result
        temp = add_polynomials(item)
        if Set(temp) != Set(globalsum)
            println(false)
            return false
        else
            continue
        end
    end
    println(true)
    return true
end

function file_parser()
    sq_matrix = []
    open("input.txt") do file
        for line in eachline(file)
            line = replace(line, r"\*" => "")

            if strip(line) != ""
                lst = split(strip(line))
                lst = [lstrip(value) for value in lst]
                push!(sq_matrix, lst)
            else
                if (length(sq_matrix) == 0)
                    continue
                end
                converted_square = extract_all(sq_matrix)
                result = magic(converted_square)
                if (result)
                    println("Magic square = True")
                else
                    println("Not a magic square = False")
                end
                sq_matrix = []
            end
        end
    end
end

file_parser()

File: Soham_Petkar.py
import re


def parser(expr):
    expr = expr.replace('^', '')
    expr = expr.replace(' ', '')
    terms = re.findall(r'([+-]?\d*)([a-z]+\d*[a-z]?\d*)?', expr)
    parsed_terms = []
    for term in terms:
        coefficient_str = term[0].strip().replace(" ", "")
        if coefficient_str == "+" or coefficient_str == "-":
            coefficient_str += "1"
        coefficient = int(coefficient_str) if coefficient_str else 1
        if expr.index(term[0]) > 0 and expr[expr.index(term[0])-1] == '-':
            coefficient *= -1
        variable = term[1]
        parsed_terms.append([coefficient, variable])
    del parsed_terms[-1]
    return parsed_terms


def add_polynomials(polynomials):
    combined_terms = {}

    for polynomial in polynomials:
        terms = parser(polynomial)
        for term in terms:
            coefficient, variable = term
            if variable in combined_terms:
                combined_terms[variable] += coefficient
            else:
                combined_terms[variable] = coefficient

    result = [(coefficient, variable)
              for variable, coefficient in combined_terms.items()]
    return result


def extract_columns(matrix):
    columns = []
    for j in range(len(matrix[0])):
        column = []
        for i in range(len(matrix)):
            column.append(matrix[i][j])
        columns.append(column)
    return columns


def extract_rows(matrix):
    rows = []
    for i in range(len(matrix)):
        rows.append(matrix[i])
    return rows


def extract_diagonals(matrix):
    diagonals = []
    diagonal = []
    # Extract main diagonal
    for i in range(len(matrix)):
        diagonal.append(matrix[i][i])
    diagonals.append(diagonal)
    diagonal = []
    # Extract secondary diagonal
    for i in range(len(matrix)):
        diagonal.append(matrix[i][len(matrix) - i - 1])
    diagonals.append(diagonal)
    return diagonals


def extract_all(matrix):
    all_items = extract_columns(
        matrix) + extract_rows(matrix) + extract_diagonals(matrix)
    return all_items


def magic(result):
    # Extract magic square
    globalsum = add_polynomials(result[0])
    for item in result:
        temp = add_polynomials(item)
        if (set(temp) != set(globalsum)):
            print(False)
            return False

        else:
            continue
    print(True)
    return True


def parse_file1(filename):
    matrices = []
    current_matrix = []

    with open(filename, 'r') as file:
        for line in file:
            line = line.strip().replace('*', '')

            if line:
                matrix_row = line.split(' ')
                current_matrix.append(matrix_row)
            else:
                if current_matrix:
                    matrices.append(current_matrix)
                current_matrix = []

        if current_matrix:
            matrices.append(current_matrix)

    return matrices


ex = parse_file1('input.txt')
for i in ex:
    print(i)
    magic(extract_all(i))
